backend/database.js

```diff
@@ -139,84 +153,84 @@ export default (sbp('sbp/selectors/register', {
     await sbp('chelonia/db/set', namespaceKey(name), value)
     return { name, value }
   },
-  'backend/db/lookupName': async function (name: string): Promise<*> {
+  'backend/db/lookupName': async function (name: string): Promise<string | Error> {
     const value = await sbp('chelonia/db/get', namespaceKey(name))
     return value || Boom.notFound()
-  },
-  // =======================
-  // Filesystem API
-  //
-  // TODO: add encryption
-  // =======================
-  'backend/db/readFile': async function (filename: string): Promise<*> {
-    const filepath = throwIfFileOutsideDataDir(filename)
-    if (!fs.existsSync(filepath)) {
-      return Boom.notFound()
-    }
-    return await readFileAsync(filepath)
-  },
-  'backend/db/writeFile': async function (filename: string, data: any): Promise<*> {
-    // TODO: check for how much space we have, and have a server setting
-    //       that determines how much of the disk space we're allowed to
-    //       use. If the size of the file would cause us to exceed this
-    //       amount, throw an exception
-    return await writeFileAsync(throwIfFileOutsideDataDir(filename), data)
-  },
-  'backend/db/writeFileOnce': async function (filename: string, data: any): Promise<*> {
-    const filepath = throwIfFileOutsideDataDir(filename)
-    if (fs.existsSync(filepath)) {
-      console.warn('writeFileOnce: exists:', filepath)
-      return
-    }
-    return await writeFileAsync(filepath, data)
   }
-}): any)
+})
+
+// Used to thwart path traversal attacks.
+export function checkKey (key: string): void {
+  if (/[/\\]/.test(key)) {
```

Comment:
One of the genuinely good ideas from that [windows filesystem link](https://learn.microsoft.com/en-us/windows/win32/fileio/naming-a-file) you mentioned is:

> Characters whose integer representations are in the range from 1 through 31,

See the unprintable ASCII characters here: https://www.rapidtables.com/code/text/ascii-table.html

I asked the bots for implementations and here are some:

```js
function hasUnprintableChars(string) {
    for (let i = 0; i < string.length; i++) {
        const charCode = string.charCodeAt(i);
        if (charCode < 32 || charCode === 127) {
            return false;
        }
    }
    return true;
}
```

```js
function containsUnprintableChars(str) {
  // Regular expression that matches any characters other than printable ASCII characters
  const unprintableCharsRegExp = /[^\x20-\x7E]/;

  return !unprintableCharsRegExp.test(str);
}
```

Just some ideas. The hard part is picking one that's fast and also accurate.

Reply:
Initially `checkKey` was only there to prevent path attacks. But if keys have to be valid Windows file names, then besides valid characters we'll also have to check for max length, reserved names, no space at the end, etc. If necessary I already have such a function lying around

Reply:
@snowteamer I think we can have a rule that is similar to but not quite the same as the Windows filesystem. I see no reason why we should ban keys with `:` in them, for example. We don't care about the reserved names like `COM1` etc. But some things really don't belong in key names, and unprintable characters are one of them.